codegen/builtin_fs.go

```diff
@@ -742,10 +769,81 @@ func (dp DockerPush) Call(ctx context.Context, cln *client.Client, ret Register,
 	return ret.Set(fs)
 }
 
-var (
-	dockerCli  *command.DockerCli
-	dockerOnce sync.Once
-)
+func pushWithMoby(ctx context.Context, dockerAPI *DockerAPIClient, ref string, l progress.SubLogger) error {
+	creds, err := imagetools.RegistryAuthForRef(ref, dockerAPI.Auth)
+	if err != nil {
+		return err
+	}
+
+	rc, err := dockerAPI.ImagePush(ctx, ref, types.ImagePushOptions{
+		RegistryAuth: creds,
+	})
+	if err != nil {
+		return err
+	}
+
+	started := map[string]*client.VertexStatus{}
+
+	defer func() {
+		for _, st := range started {
+			if st.Completed == nil {
+				now := time.Now()
+				st.Completed = &now
+				l.SetStatus(st)
+			}
+		}
+	}()
+
+	dec := json.NewDecoder(rc)
+	var parsedError error
+	for {
+		var jm jsonmessage.JSONMessage
+		if err := dec.Decode(&jm); err != nil {
+			if parsedError != nil {
+				return parsedError
+			}
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if jm.ID != "" {
+			id := "pushing layer " + jm.ID
+			st, ok := started[id]
+			if !ok {
+				if jm.Progress != nil || jm.Status == "Pushed" {
+					now := time.Now()
+					st = &client.VertexStatus{
+						ID:      id,
+						Started: &now,
+					}
+					started[id] = st
+				} else {
+					continue
+				}
+			}
+			st.Timestamp = time.Now()
+			if jm.Progress != nil {
+				st.Current = jm.Progress.Current
+				st.Total = jm.Progress.Total
+			}
+			if jm.Error != nil {
+				now := time.Now()
+				st.Completed = &now
+			}
+			if jm.Status == "Pushed" {
+				now := time.Now()
+				st.Completed = &now
+				st.Current = st.Total
+			}
+			l.SetStatus(st)
+		}
+		if jm.Error != nil {
+			parsedError = jm.Error
```

Comment:
Why do we save this instead of returning it right away?

Reply:
Tbh, I'm not too sure, I ported this code from a buildx private function. It looks like even if there is a json message error, we want to continue as long as the next decode is not also an error (io.EOF, or others).

I can experiment with exiting early but I'm not familiar with moby/moby's docker push progress code.

Reply:
I think it so the code continues reading the `rc` stream until `EOF`.  It will return the parsedError only after `dec.Decode` has an error, which is likely `EOF`.